Use a non-empty, non-'about:' cookie url when getting/setting cookies in CookieStore

RupinMittal opened the pull request:
#### 4973ae3a5e854e1e056dedcbced9c1db450bc6f0
<pre>
Use a non-empty, non-&apos;about:&apos; cookie url when getting/setting cookies in CookieStore
<a href="https://bugs.webkit.org/show_bug.cgi?id=279578">https://bugs.webkit.org/show_bug.cgi?id=279578</a>
<a href="https://rdar.apple.com/135853069">rdar://135853069</a>

Reviewed by Chris Dumez and Sihui Liu.

As per the CookieStore API spec, the CookieStore uses the ScriptExecutionContext&apos;s
creation URL when getting/setting cookies. Two of the layout tests attempt to get/set
cookies from iFrames. The creation URL&apos;s for these iFrames are &apos;about:blank&apos; and
&apos;about:srcdoc&apos;. These url&apos;s don&apos;t have hosts and so the get/set requests fail.

We fix this by using Document::cookieURL() in the Document case. The document&apos;s cookie
URL is set when the Document is created and never changed after that (so it matches the
creation URL). This works in the iFrame case because Document::initSecurityContext()
sets the cookieURL to the parent document&apos;s cookies URL for a Document if it
shouldInheritSecurityOriginFromOwner (which is true for iFrames).

We added ScriptExecutionContext::creationURL() in 283530@main specifically for getting/
setting cookies. Since Document had a creationURL() function, we wanted one in
ServiceWorkerGlobalScope so that the usage in CookieStore was consistent. Since we are
now using cookieURL() in Document instead of creationURL(), we want to do the same in
ServiceWorkerGlobalScope. So we rename ScriptExecutionContext::creationURL() to
ScriptExecutionContext::cookieURL(). So now, we use cookieURL() in both the Document
and Service Worker case. This works in all cases, not just for iFrames.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_across_frames.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_opaque_origin.https-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::get):
(WebCore::CookieStore::getAll):
(WebCore::CookieStore::set):
* Source/WebCore/dom/Document.h:
(WebCore::Document::creationURL const):
(WebCore::Document::cookieURL const): Deleted.
* Source/WebCore/dom/EmptyScriptExecutionContext.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebCore/worklets/WorkletGlobalScope.h:

Canonical link: <a href="https://commits.webkit.org/283622@main">https://commits.webkit.org/283622@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21b0dc39419bf5119304ae9f050b619c22589795

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66782 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46157 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19404 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70817 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17915 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68900 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53956 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17731 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53504 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12060 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69849 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42473 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57774 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34139 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39145 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15166 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16269 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61059 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15507 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72518 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10739 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14898 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60912 "Found 1 new test failure: fast/css/focus-ring-exists-for-search-field.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10771 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57831 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61147 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8824 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2436 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10141 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41964 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43041 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44224 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42784 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -230,6 +231,20 @@ CookieStore::~CookieStore()
 
+URL CookieStore::creationURL(ScriptExecutionContext& context) const
+{
+    auto url = context.creationURL();
```
I think `document.cookieURL()` gives the right value

RupinMittal replied:
I believe `document.cookieURL()` won't work here. The problem with using `document.url()` is that this returns the document's current URL (`m_url` in Document.h) and this can change with functions like `history.pushState`. And  `m_cookieURL` (which is what `document.cookieURL()` returns) can be set with `m_url`. For example: `setCookieURL(m_url)` in `Document::initSecurityContext()`. So I imagine it's probably best to use the creation URL explicitly as the spec says.

cdumez replied:
Can you try cookieURL and see if it works? This is what WebKit is using for cookies elsewhere so being inconsistent here would be unfortunate (if it can be avoided).

szewai replied:
Did you test? `Document::initSecurityContext()` should set cookie URL to set parent's url if `shouldInheritSecurityOriginFromOwner`: https://github.com/WebKit/WebKit/blob/280eaf1c4a30cd0650835a410435655ad244f297/Source/WebCore/dom/Document.cpp#L7651

szewai replied:
Also, I think the new API must use what `document.cookie` is using now, otherwise it's a leak.

RupinMittal replied:
That's a good point. I hadn't tried it yet (oops)--but I just did, and it works! It looks much cleaner and more consistent this way. Thank you!

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -230,6 +231,20 @@ CookieStore::~CookieStore()
     m_mainThreadBridge->detach();
 }
 
+URL CookieStore::creationURL(ScriptExecutionContext& context) const
```
Seems like it should be a static function in the cpp. It doesn't seem to rely on any class member?

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -230,6 +231,20 @@ CookieStore::~CookieStore()
+URL CookieStore::creationURL(ScriptExecutionContext& context) const
+{
+    auto url = context.creationURL();
+    if (context.isDocument()) {
```
`if (RefPtr document = dynamicDowncast<Document>(context)) {`

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -230,6 +231,20 @@ CookieStore::~CookieStore()
+{
+    auto url = context.creationURL();
+    if (context.isDocument()) {
+        RefPtr document = dynamicDowncast<Document>(context);
```
Not needed after change above.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -230,6 +231,20 @@ CookieStore::~CookieStore()
+        RefPtr document = dynamicDowncast<Document>(context);
+        RefPtr parentDocument = document->protectedParentDocument();
+        while (SecurityPolicy::shouldInheritSecurityOriginFromOwner(url)) {
+            url = parentDocument->creationURL();
```
What if parentDocument is null?

RupinMittal replied:
Good point--we should throw an error if we can't find a creation URL. Just added this in the new iteration.

RupinMittal replied:
As suggested by Sihui, I changed it to use Document::cookieURL(), so now we don't access the parentDocument anymore.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -230,6 +230,13 @@ CookieStore::~CookieStore()
     m_mainThreadBridge->detach();
 }
 
+static URL creationURL(ScriptExecutionContext& context)
```
Can probably use `const ScriptExecutionContext&`

RupinMittal replied:
Changed to using cookieURL() everywhere, so we don't need this function anymore.

szewai commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -230,6 +230,13 @@ CookieStore::~CookieStore()
+{
+    if (RefPtr document = dynamicDowncast<Document>(context))
+        return document->cookieURL();
+    return context.creationURL();
```
We should probably just rename `ScriptExecutionContext:: creationURL()` to `cookieURL()`, then we don't need to check here.
The [creation URL](https://html.spec.whatwg.org/multipage/webappapis.html#concept-environment-creation-url) is represented by `ScriptExecutionContext::url()` for non-document context.

RupinMittal replied:
That's a good suggestion--we added `ScriptExecutionContext:: creationURL()` specifically so that ServiceWorkerGlobalScope would have a `creationURL()` function just like Document to be consistent for getting the url needed for cookies (and it just used `ScriptExecutionContext::url()` under the hood). Now that Document uses `cookieURL()`, it's better to do the same for ServiceWorkerGlobalScope too. I've renamed it `ScriptExecutionContext::cookieURL()`.

RupinMittal committed:
Use a non-empty, non-'about:' cookie url when getting/setting cookies in CookieStore
https://bugs.webkit.org/show_bug.cgi?id=279578
rdar://135853069

Reviewed by Chris Dumez and Sihui Liu.

As per the CookieStore API spec, the CookieStore uses the ScriptExecutionContext's
creation URL when getting/setting cookies. Two of the layout tests attempt to get/set
cookies from iFrames. The creation URL's for these iFrames are 'about:blank' and
'about:srcdoc'. These url's don't have hosts and so the get/set requests fail.

We fix this by using Document::cookieURL() in the Document case. The document's cookie
URL is set when the Document is created and never changed after that (so it matches the
creation URL). This works in the iFrame case because Document::initSecurityContext()
sets the cookieURL to the parent document's cookies URL for a Document if it
shouldInheritSecurityOriginFromOwner (which is true for iFrames).

We added ScriptExecutionContext::creationURL() in 283530@main specifically for getting/
setting cookies. Since Document had a creationURL() function, we wanted one in
ServiceWorkerGlobalScope so that the usage in CookieStore was consistent. Since we are
now using cookieURL() in Document instead of creationURL(), we want to do the same in
ServiceWorkerGlobalScope. So we rename ScriptExecutionContext::creationURL() to
ScriptExecutionContext::cookieURL(). So now, we use cookieURL() in both the Document
and Service Worker case. This works in all cases, not just for iFrames.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_across_frames.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_opaque_origin.https-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::get):
(WebCore::CookieStore::getAll):
(WebCore::CookieStore::set):
* Source/WebCore/dom/Document.h:
(WebCore::Document::creationURL const):
(WebCore::Document::cookieURL const): Deleted.
* Source/WebCore/dom/EmptyScriptExecutionContext.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebCore/worklets/WorkletGlobalScope.h:

Canonical link: https://commits.webkit.org/283622@main
